include/nbl/video/IGPUCommandPool.h

```diff
@@ -4,38 +4,835 @@
+
+            cmdMem = nextSegment->allocate<Cmd>(args...);
+            if (!cmdMem)
+            {
+                assert(false);
+                return nullptr;
+            }
+
+            segmentListTail->setNext(nextSegment);
+            segmentListTail = segmentListTail->getNext();
+        }
+        Cmd* cmd = new (cmdMem) Cmd(std::forward<Args>(args)...);
+
+        if (segmentListHeadItr.m_cmd == nullptr)
+            segmentListHeadItr.m_cmd = cmd;
+
+        return cmd;
+    }
+
+    void deleteCommandSegmentList(CCommandSegment::Iterator& segmentListHeadItr, CCommandSegment*& segmentListTail)
+    {
+        auto& itr = segmentListHeadItr;
+
+        if (itr.m_segment && itr.m_cmd)
+        {
+            bool lastCmd = itr.m_cmd->getSize() == 0u;
+            while (!lastCmd)
+            {
+                ICommand* currCmd = itr.m_cmd;
+                CCommandSegment* currSegment = itr.m_segment;
+
+                itr.m_cmd = reinterpret_cast<ICommand*>(reinterpret_cast<uint8_t*>(itr.m_cmd) + currCmd->getSize());
+                currCmd->~ICommand();
+                // No need to deallocate currCmd because it has been allocated from the LinearAddressAllocator where deallocate is a No-OP and the memory will
+                // get reclaimed in ~LinearAddressAllocator
+
+                // We potentially continue to the next command segment under any one of the two conditions:
+                const bool potentiallyContinueToNextSegment =
+                    // 1. If the we run past the storage of the current segment.
+                    ((reinterpret_cast<uint8_t*>(itr.m_cmd) - itr.m_segment->getData()) >= IGPUCommandPool::CCommandSegment::STORAGE_SIZE)
+                    ||
+                    // 2. If we encounter a 0-sized command (terminating command) before running out of the current segment. This case will arise when the current
+                    // segment doesn't have enough storage to hold the next command.
+                    (itr.m_cmd->getSize() == 0);
+
+                if (potentiallyContinueToNextSegment)
+                {
+                    CCommandSegment* nextSegment = currSegment->getNext();
+
+                    currSegment->~CCommandSegment();
+                    m_commandSegmentPool.deallocate(currSegment, COMMAND_SEGMENT_SIZE);
+
+                    if (!nextSegment)
+                        break;
+
+                    itr.m_segment = nextSegment;
+                    itr.m_cmd = itr.m_segment->getFirstCommand();
+                }
+
+                lastCmd = itr.m_cmd->getSize() == 0u;
+                if (lastCmd)
+                {
+                    currSegment->~CCommandSegment();
+                    m_commandSegmentPool.deallocate(currSegment, COMMAND_SEGMENT_SIZE);
+                }
+            }
+
+            segmentListHeadItr.m_cmd = nullptr;
+            segmentListHeadItr.m_segment = nullptr;
+            segmentListTail = nullptr;
```

Comment:
make the assignments uncoditional, saves us fuckups in the future should the `if` change

Reply:
Resolved.